testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java

```diff
@@ -142,15 +146,107 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
 
     @Test
     @Timeout(value = 10000, unit = TimeUnit.MILLISECONDS)
-    public void testCompositeBufferPartialWriteDoesNotCorruptData(TestInfo testInfo) throws Throwable {
-        run(testInfo, this::testCompositeBufferPartialWriteDoesNotCorruptData);
+    public void testSingleCompositeBufferWrite(TestInfo testInfo) throws Throwable {
+        run(testInfo, this::testSingleCompositeBufferWrite);
     }
 
-    protected void compositeBufferPartialWriteDoesNotCorruptDataInitServerConfig(ChannelConfig config,
-                                                                                 int soSndBuf) {
+    public void testSingleCompositeBufferWrite(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+        enableNewBufferAPI(sb, cb);
+        Channel serverChannel = null;
+        Channel clientChannel = null;
+        try {
+            final CountDownLatch latch = new CountDownLatch(1);
+            final AtomicReference<Object> clientReceived = new AtomicReference<>();
+            sb.childHandler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    ch.pipeline().addLast(new ChannelHandler() {
+                        @Override
+                        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                            ctx.writeAndFlush(newCompositeBuffer(ctx.bufferAllocator()))
+                                    .addListener(ctx, ChannelFutureListeners.CLOSE);
+                        }
+                    });
+                }
+            });
+            cb.handler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    ch.pipeline().addLast(new ChannelHandler() {
+                        private Buffer aggregator;
+                        @Override
+                        public void handlerAdded(ChannelHandlerContext ctx) {
+                            aggregator = ctx.bufferAllocator().allocate(EXPECTED_BYTES);
+                        }
+
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                            if (msg instanceof Buffer) {
+                                try (Buffer buf = (Buffer) msg) {
+                                    aggregator.writeBytes(buf);
+                                }
+                            }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
```

Comment:
Should we also close the `aggregator` in this case ?

Reply:
Does `channelInactive` not take care of it regardless?